lib/_http_outgoing.js

```diff
@@ -198,57 +181,72 @@ function _storeHeader(firstLine, headers) {
   // firstLine in the case of request is: 'GET /index.html HTTP/1.1\r\n'
   // in the case of response it is: 'HTTP/1.1 200 OK\r\n'
   var state = {
-    sentConnectionHeader: false,
-    sentConnectionUpgrade: false,
-    sentContentLengthHeader: false,
-    sentTransferEncodingHeader: false,
-    sentDateHeader: false,
-    sentExpect: false,
-    sentTrailer: false,
-    sentUpgrade: false,
-    messageHeader: firstLine
+    connection: false,
+    connUpgrade: false,
+    contLen: false,
+    te: false,
+    date: false,
+    expect: false,
+    trailer: false,
+    upgrade: false,
+    header: firstLine
   };
 
-  var i;
-  var j;
   var field;
+  var key;
   var value;
-  if (headers instanceof Array) {
-    for (i = 0; i < headers.length; ++i) {
+  var i;
+  var j;
+  if (headers === this._headers) {
+    for (key in headers) {
+      var entry = headers[key];
+      field = entry[0];
+      value = entry[1];
+
+      if (value instanceof Array) {
```

Comment:
It should be, since we're the ones storing header values in `this._headers`. That's also the reason why we can use for-in instead of `Object.keys()` (slight speedup there) when iterating over `this._headers`.